### PR TITLE
Make libnotify depdency optional

### PR DESCRIPTION
**Pull request by ChillerDragon**
Compile the client without notification support instead of throwing a compile error.
Shows a warning in the cmake step if it does not find libnotify.

image

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
What's the motivation for this pull request? Why can't you just install libnotify?

**Comment by ChillerDragon** [comment]
> What's the motivation for this pull request? Why can't you just install libnotify?

I would like to compile a headless ddnet client in minimal linux distributions such as alpine

https://pkgs.alpinelinux.org/packages?name=notify&branch=edge&repo=&arch=&maintainer=

**Comment by heinrich5991** [comment]
> Also I am always super happy when I checkout new projects which make compiling easier by allowing optional dependencies.

I don't think anyone on a normal distribution doesn't have libnotify installed.

> I would like to compile a headless ddnet client in minimal linux distributions such as alpine

Why would you like to do this?

(I'm asking these questions because you're asking DDNet to take over maintenance for this use case. I'm not sure this is a common use case.)

**Comment by ChillerDragon** [comment]
> I don't think anyone on a normal distribution doesn't have libnotify installed.

Agreed. If you don't want to make the life easier for other distros due to maintenance that is fair i guess.

> Why would you like to do this?

To get a small sized container image. Like for example the one I published for the [server](https://hub.docker.com/r/chillerdragon/ddnet-mysql).

**Comment by heinrich5991** [comment]
> To get a small sized container image. Like for example the one I published for the [server](https://hub.docker.com/r/chillerdragon/ddnet-mysql).

What use case do you have for this? I don't know what use the headless client is, other than for integration tests, e.g. ^^ It is very useful for that and I'm thankful that you pushed us to include that!

File: CMakeLists.txt [hunk 1]
```diff
@@ -775,6 +775,12 @@ else()
   else()
     set(PLATFORM_LIBS)
   endif()
+
+  if(NOT(NOTIFY_FOUND))
+    message(WARNING "libnotify not found compiling without notification support")
```
[the remark this answers is not in the document]
Comment (ChillerDragon) [reply CMakeLists.txt]: english pro :england: 
thank

**Comment by heinrich5991** [comment]
> Imo the headless client is the `curl` equivalent for the ddnet protocol. Its nice for automation. And testing connectivity. Helps debugging firewall and similar issues when you can quickly spin up a client on the same network that the server is.

Is that something you do? When is that useful? If it's just a hypothetical scenario, I don't think I want to change DDNet's code for it. Also, the way it's currently implemented hurts reproducible builds because the artifact depends on the system configuration.

**Comment by ChillerDragon** [comment]
> > Imo the headless client is the `curl` equivalent for the ddnet protocol. Its nice for automation. And testing connectivity. Helps debugging firewall and similar issues when you can quickly spin up a client on the same network that the server is.
> 
> Is that something you do? When is that useful? If it's just a hypothetical scenario, I don't think I want to change DDNet's code for it. Also, the way it's currently implemented hurts reproducible builds because the artifact depends on the system configuration.

I have used headless clients to test firewalls/port bound issues etc already. For example when I can not reach ger10 from my home network I sometimes spin up a headless client on my vps to see if it works. Or if I can not connect to a server where I have ssh access to I try connecting from localhost to see if it is my a issue with the server code and not the firewall and so on. The containerized image could be nice to quickly spin up a small test pod in a kubernetes environment to test the routing setup there.

But no hard feelings on this pr. Feel free to close it if it concerns you. Libnotify wouldn't fix everything. SDL2 is probably another problem as well at least for alpine. When someone wants to run ddnet in serenityOS [SDL2](https://github.com/SerenityOS/serenity/tree/master/Ports/SDL2) and [freetype](https://github.com/SerenityOS/serenity/tree/master/Ports/freetype) would be available while libnotify would not. Also getting libnotify installed on [openSUSE Leap](https://software.opensuse.org/search?baseproject=openSUSE%3ALeap%3A15.5&q=notify) is a non obvious task for me. I personally think that if one tries to test ddnet on a distro like openSUSE every annoying dependency that enables something so gameplay irrelevant such as desktop notification brings a potential player closer to rage quitting during the first compile.

**Comment by heinrich5991** [comment]
That sounds like a sane use case to me. Have you tested that this still includes the libnotify code when libnotify is found?

> For example when I can not reach ger10 from my home network I sometimes spin up a headless client on my vps to see if it works.

Could probably be done more cheaply by reimplementing the functionality.

**Comment by ChillerDragon** [comment]
> That sounds like a sane use case to me. Have you tested that this still includes the libnotify code when libnotify is found?

I tested this on debian by apt removing and adding libnotify.
I tested that the build passes without and that the popups show in gnome when libnotify is installed and one pings you in game.

**Comment by heinrich5991** [comment]
If this were to be added, it should be a configuration option like `VULKAN` instead of autodetection. All in all, I think it shouldn't be added until we come across a used system where this is actually a problem. See also #6893.

**Pull request closed without merging**